testutil/promrated/rated.go

```diff
@@ -0,0 +1,125 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package promrated
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/expbackoff"
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/app/z"
+)
+
+type validatorEffectivenessData struct {
+	Uptime                 float64 `json:"uptime"`
+	AvgCorrectness         float64 `json:"avgCorrectness"`
+	AvgInclusionDelay      float64 `json:"avgInclusionDelay"`
+	AttesterEffectiveness  float64 `json:"attesterEffectiveness"`
+	ProposerEffectiveness  float64 `json:"proposerEffectiveness"`
+	ValidatorEffectiveness float64 `json:"validatorEffectiveness"`
+}
+
+// getValidationStatistics queries rated for a pubkey and returns rated data about the pubkey
+func getValidationStatistics(ctx context.Context, ratedEndpoint string, validator validator) (*validatorEffectivenessData, error) {
+	url, err := url.Parse(ratedEndpoint)
+	if err != nil {
+		return nil, errors.Wrap(err, "parse rated endpoint")
+	}
+
+	url.Path = fmt.Sprintf("/v0/eth/validators/%s/effectiveness", validator.PubKey)
+
+	query := url.Query()
+	query.Add("size", "1")
+	url.RawQuery = query.Encode()
+
+	maxRetries := 10
+
+	client := new(http.Client)
+
+	for r := 0; r <= maxRetries; r++ {
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
```

Comment:
always define timeouts when doing IO: https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/